### PR TITLE
Add flop counter utility

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95901
* #95854
* __->__ #95751
* #95355

Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
image

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

image

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]
FlopCounterMode.register_hierarchy(f, name): Allows you to register additional "hierarchies" for a function.

### CODE REVIEW

**Comment by ezyang** [comment]
Who do you want to do a close code review on this?

File: torch/utils/flop_counter.py
Comment (yinghai) [inline torch/utils/flop_counter.py:134]: Can we also add support to scaled_dot_product_attention since it's more and more used nowadays?

Comment (Chillee) [reply torch/utils/flop_counter.py:134]: Good idea.

**Comment by Chillee** [comment]
@ezyang Not sure I really need a close code review from anybody 🤔 Maybe @ngimel

File: torch/utils/flop_counter.py
Comment (ppwwyyxx) [inline torch/utils/flop_counter.py:301]: Is it a good idea to always apply get_shape on tensors? For example I wonder if there would be an op that takes a scalar boolean tensor and condition two different behaviors (with different flops) on it.

File: torch/utils/flop_counter.py
Comment (ppwwyyxx) [reply torch/utils/flop_counter.py:134]: And einsum

Comment (Chillee) [reply torch/utils/flop_counter.py:134]: Einsum isn’t needed since it’s a “compositeimplicit” op and decomposes into other operators.

File: test/test_flop_counter.py
Comment (lezcano) [inline test/test_flop_counter.py:41]: Odd. 4 * 6 * 2 * 5 = 240. But the number of FLOPS for an `mm` `[m, k] x [k, n]` is `m * n * (2 * k - 1)`, which in this case gives 216.

File: torch/utils/flop_counter.py
Comment (lezcano) [inline torch/utils/flop_counter.py:33]: Why not doing it right, now that we're at it?

File: torch/testing/_internal/common_cuda.py
Comment (albanD) [inline torch/testing/_internal/common_cuda.py:30]: This needs to be reverted before merging.

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:150]: ?

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:258]:
```suggestion
        depth = depth or self.depth or 999999
```

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:241]: What is this doing?

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:185]: You don't actually need the user to pass in the module or use custom Function.
You can use global hooks like `torch.nn.modules.module.register_module_forward_pre_hook()` that will be called for ALL modules.

More generally, you can use `forward_pre`/`forward` hook to do the forward push-pop and the `full_backward_pre_hook`/`full_backward_hook` to do the backward push pop.

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:118]: Should you just pass in the args/kwargs to these functions? That will simplify the logic in here quite a bit no?

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:15]: from math import prod

**Review by albanD — COMMENTED** [review]
There are a lot of public not documented functions ;)

File: torch/utils/flop_counter.py
Comment (ppwwyyxx) [reply torch/utils/flop_counter.py:185]: I think that will be a good API improvement. But in that case, what shall we do if users call two different models (which might have shared or distinct layers with the same name) in the context? This might get tricky to handle.

File: test/test_flop_counter.py
Comment (ppwwyyxx) [inline test/test_flop_counter.py:19]: It seems the test would be simpler if this function just returns int.

File: test/test_flop_counter.py
Comment (ppwwyyxx) [inline test/test_flop_counter.py:120]: Other cases to test:
* Module A call a submodule `self.subm.method_name()`, instead of `self.subm.__call__`.  It's ambiguous in this case which module (self or self.subm) the flops should attribute to, but please test that the results are at least reasonable.
* Module with scripted submodule: https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/tests/test_flop_count.py#L832-L845
* Module with legacy autograd function: https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/tests/test_flop_count.py#L820-L830  Not sure if this could be supported at all.

File: torch/utils/flop_counter.py
Comment (ppwwyyxx) [inline torch/utils/flop_counter.py:288]: maybe `return self`? so users can do:
```
with FlopCountingMode(...) as flop:
   do stuff
```

File: torch/testing/_internal/common_cuda.py
Comment (Chillee) [reply torch/testing/_internal/common_cuda.py:30]: oops

File: torch/utils/flop_counter.py
Comment (ppwwyyxx) [inline torch/utils/flop_counter.py:179]: Users want to access this, either for a custom display format, or for aggregating it across a dataset (e.g. https://github.com/facebookresearch/detectron2/blob/38af375052d3ae7331141bc1a22cfa2713b02987/tools/analyze_model.py#L66-L68)

So it's probably better to make it a proper API with documentation.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:301]: I don't think it's a big deal - if that ever shows up we can just pass through scalar tensors.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:241]: Hmm... it's just providing a way for users to register a custom level in the hierarchy. Perhaps it's unneeded tbh - might remove it.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:118]: Hmm... maybe.

File: test/test_flop_counter.py
Comment (Chillee) [reply test/test_flop_counter.py:19]: assertExpectedInline expects strings.

File: test/test_flop_counter.py
Comment (Chillee) [reply test/test_flop_counter.py:120]: > Module with scripted submodule:

Unfortunately we can't support hierarchies on scripted modules, since they don't support hooks. I'll add a test for just capturing flops under scripted functions.

> Module with legacy autograd function:

This works totally fine.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:185]: oh, full_backward_pre_hook is a new API :P

Comment (Chillee) [reply torch/utils/flop_counter.py:185]: In general, I don't think global module hooks make that much sense given that we're trying to register a hierarchy. Concretely, the hook doesn't tell us where in the module hierarchy it lives.

Comment (Chillee) [reply torch/utils/flop_counter.py:185]: Hmm... I needed to do the "clone" thing in order to work around this issu
```
RuntimeError: Output 0 of BackwardHookFunctionBackward is a view and is being modified inplace. This view was created inside a custom Function (or because an input was returned as-is) and the autograd logic to handle view+inplace would override the custom backward associated with the custom Function, leading to incorrect gradients. This behavior is forbidden. You can fix this by cloning the output of the custom Function.
```

How would I do this with `full_backward_pre_hook`?

File: test/test_flop_counter.py
Comment (ngimel) [inline test/test_flop_counter.py:71]: it does, via output_size. To compute each output value, you need `kernel_size * c_in * c_out` flops. Sometimes in padded cases you could do super inefficient implementation that computes and throws away padded values, but roofline flop counter shouldn't account for implementation inefficiencies

File: torch/utils/flop_counter.py
Comment (ngimel) [inline torch/utils/flop_counter.py:44]: wait, why? `mm` flops already account for this?

File: torch/utils/flop_counter.py
Comment (ngimel) [inline torch/utils/flop_counter.py:85]: what addition?

File: test/test_flop_counter.py
Comment (ngimel) [inline test/test_flop_counter.py:49]: add tests with `alpha` and `beta` args

File: torch/utils/flop_counter.py
Comment (ngimel) [inline torch/utils/flop_counter.py:301]: Non-tensor args for addmm/baddbmm could throw off flop_count functions I think

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:258]: this doesn't work if depth is 0.

File: test/test_flop_counter.py
Comment (Chillee) [reply test/test_flop_counter.py:71]: Yeah, but doesn't the padding increase output_size even though we don't need to actually compute them?

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:44]: I was just adding the addition flops, although maybe this is unneeded 🤔

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:301]: Since if they're 0 we don't actually do the computation?

Comment (ngimel) [reply torch/utils/flop_counter.py:301]: But `get_shape` would still return them unchanged, and then `addmm_flops` will have 5 args instead of 3?

Comment (Chillee) [reply torch/utils/flop_counter.py:301]: I think those are always passed in as kw-only args.

Comment (ngimel) [reply torch/utils/flop_counter.py:301]: Ah cool, yeah docs say kw-only

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py]: Decided to remove them for consistency.

File: test/test_flop_counter.py
Comment (ngimel) [reply test/test_flop_counter.py:72]: I guess you could claim that, but in reality you are still computing them in implicit-gemm formulation, those 0's are still part of the gemm tile.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:121]: I'm gonna do the attention ones in a follow-up PR.

**Review by ngimel — APPROVED** [review]

File: torch/utils/flop_counter.py
Comment (ppwwyyxx) [reply torch/utils/flop_counter.py:121]: This actually leads to an important design issue: I think there should be an API to return "unsupported" ops. Similar to https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/fvcore/nn/jit_analysis.py#L98

The reason is that users are mainly interacting with high-level modules and ops, and aren't aware of what low-level ops are called. Imagine one day torch adds a new op for transformer, or adds a different implementation of einsum that doesn't decompose: the change should be transparent, but it makes the flop counter become a silent error!

So in fvcore we provide unsupported ops and print a warning about them by default. To make the results more meaningful, we have a list of trivial ops that are always ignored and will not appear in "unsupported ops" (https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/fvcore/nn/jit_analysis.py#L28).

Comment (Chillee) [reply torch/utils/flop_counter.py:121]: @ppwwyyxx Hmm... I think I might just add all the ops as "unsupported ops". 

I think one benefit of putting it in core is that in principle, we should hopefully be able to keep the operator list up to date better.

Comment (ngimel) [reply torch/utils/flop_counter.py:121]: Yeah, impl changes should be caught by tests, we have an einsum test in this PR, but not matmul?

File: torch/utils/flop_counter.py
Comment (albanD) [reply torch/utils/flop_counter.py]: > Concretely, the hook doesn't tell us where in the module hierarchy it lives.

But you can use a global stack to track where you are in the hierarchy right now. Like you know what the current Module is and when you enter a new one, make the current its parent and set the new one as the new current.

> Hmm... I needed to do the "clone" thing in order to work around this issu

I'm not sure what you mean by the clone thing. But yes if you have custom Function that is an issue.
If you have time, I think we can rework this to use multi-Tensor hooks instead of custom Function and that will remove this constraint :)

File: torch/utils/flop_counter.py
Comment (albanD) [inline torch/utils/flop_counter.py:131]: Please document public API or make them private if they shouldn't be public.

**Review by albanD — CHANGES_REQUESTED** [review]
I need to upgrade the public API script again to pick these up...

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:131]: ah that's what you meant

Comment (Chillee) [reply torch/utils/flop_counter.py:131]: I can also just set `__all__` right?

Comment (albanD) [reply torch/utils/flop_counter.py:133]: Yes but then you'll see that you can't have these functions that look public and are not in `__all__`. So you will have to prepend them with `_` anyways.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py]: If I have a module like `ReLU` where the forwards pass is
```
def forward(self, x):
    return x.relu_()
```

And I install a backwards hook, it throws that error.

File: torch/utils/flop_counter.py
Comment (Chillee) [reply torch/utils/flop_counter.py:133]: Why will I see? It seems to work for me.

**Review by albanD — APPROVED** [review]
Ok if CI is ok!

File: torch/utils/flop_counter.py
Comment (danthe3rd) [inline torch/utils/flop_counter.py:214]: quick question: why are we calling `clone()` here?

Comment (Chillee) [reply torch/utils/flop_counter.py:214]: If the output of an autograd.Function is a view of the input, there are some restrictions on whether you can use inplace operations or not.

File: test/test_flop_counter.py
Comment (ppwwyyxx) [reply test/test_flop_counter.py:157]: Actually, today I just found that autograd function isn't working in the way I expected.

Usually users will put custom cuda kernels, which pytorch knows nothing about, in such autograd function. For example like this https://github.com/facebookresearch/detectron2/blob/e9f7e2ba15abd7badcb05ef6f5076f06b36a9c5b/detectron2/layers/deform_conv.py#L59

It seems in this PR there is no way to register a custom flop counter for custom kernels that are not registered as a pytorch op (e.g. https://github.com/facebookresearch/detectron2/blob/e9f7e2ba15abd7badcb05ef6f5076f06b36a9c5b/detectron2/layers/csrc/vision.cpp#L83-L94).

However, the fvcore flop counter can interpret the autograd function as an atomic unit (instead of entering it, like this PR), so it actually allows users to count custom kernel properly, for example in this test: https://github.com/facebookresearch/fvcore/blob/e8d19df2ffdaf0a78c8d88d6a2522c36b0cacb07/tests/test_flop_count.py#L820-L830

**Pull request closed without merging**
